### PR TITLE
Feature (library): BYO client and dialer

### PR DESCRIPTION
**Pull request by sjauld**
#### Background

Temp solution for #263 whilst #305 is under development

#### Checklist

- [/] Git commit messages conform to [community standards](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [/] Each Git commit represents meaningful milestones or atomic units of work.
- [/] Changed or added code is covered by appropriate tests.

### CODE REVIEW

File: lib/attack.go [hunk 1]
```diff
@@ -204,6 +204,16 @@ func MaxBody(n int64) func(*Attacker) {
 	return func(a *Attacker) { a.maxBody = n }
 }
 
+// Client returns a functional option that allows you to BYO http.Client
```
Comment (tsenart) [inline lib/attack.go]: BYO is unnecessarily abbreviated and may lead to some users needing to lookup the term. Let's err on the side of explicitness here.

File: lib/attack.go [hunk 2]
```diff
@@ -204,6 +204,16 @@ func MaxBody(n int64) func(*Attacker) {
...
+func Client(c http.Client) func(*Attacker) {
+	return func(a *Attacker) { a.client = c }
+}
+
+// Dialer returns a functional option that allows you to BYO *net.Dialer
+func Dialer(d *net.Dialer) func(*Attacker) {
```
Comment (tsenart) [inline lib/attack.go]: Given we already allow setting the `http.Client` above, do we need this option? You can set the dialer in the client directly.

File: lib/attack.go [hunk 3]
```diff
@@ -204,6 +204,16 @@ func MaxBody(n int64) func(*Attacker) {
 	return func(a *Attacker) { a.maxBody = n }
 }
 
+// Client returns a functional option that allows you to BYO http.Client
+func Client(c http.Client) func(*Attacker) {
```
Comment (tsenart) [inline lib/attack.go]: I think we should pass a pointer here unless you have a specific reason not to.

File: lib/attack_test.go [hunk 4]
```diff
@@ -266,3 +267,73 @@ func TestMaxBody(t *testing.T) {
...
+	tr := NewStaticTargeter(Target{Method: "GET", URL: server.URL})
+	rate := Rate{Freq: 100, Per: time.Second}
+
+	atk := NewAttacker(Client(client))
+	var hits uint64
+	for range atk.Attack(tr, rate, 1*time.Second, "") {
+		hits++
+	}
+
+	if got, want := hits, uint64(rate.Freq); got != want {
```
Comment (tsenart) [inline lib/attack_test.go]: This doesn't actually test that a specific client was used. Perhaps an easy way to do so would be to set a timeout of zero and check for the particular error is returned.

**Comment by sjauld** [comment]
Changes implemented!

**Comment by tsenart** [comment]
Nice, thanks!

**Pull request merged**